nvflare/fuel/f3/drivers/net_utils.py

```diff
@@ -0,0 +1,282 @@
+from ssl import SSLContext
+from typing import Any, Optional
+from urllib.parse import parse_qsl, urlencode, urlparse
+
+from nvflare.fuel.f3.comm_error import CommError
+from nvflare.fuel.f3.drivers.driver_params import DriverParams
+
+log = logging.getLogger(__name__)
+
+LO_PORT = 1025
+HI_PORT = 65535
+MAX_ITER_SIZE = 10
+RANDOM_TRIES = 20
+SECURE_SCHEMES = {"https", "wss", "grpcs", "stcp", "satcp"}
+
+
+def bool_value(value) -> bool:
+    """Convert value into bool"""
+    if isinstance(value, str):
+        return value.lower() in {"true", "yes", "y", "t", "1"}
+
+    return value
+
+
+def ssl_required(params: dict) -> bool:
+    """Check if SSL is required"""
+    scheme = params.get(DriverParams.SCHEME.value, None)
+    return scheme in SECURE_SCHEMES or bool_value(params.get(DriverParams.SECURE.value, False))
+
+
+def get_ssl_context(params: dict, ssl_server: bool) -> Optional[SSLContext]:
+    if not ssl_required(params):
+        return None
+
+    ca_path = params.get(DriverParams.CA_CERT.value)
+    if ssl_server:
+        cert_path = params.get(DriverParams.SERVER_CERT.value)
+        key_path = params.get(DriverParams.SERVER_KEY.value)
+    else:
+        cert_path = params.get(DriverParams.CLIENT_CERT.value)
+        key_path = params.get(DriverParams.CLIENT_KEY.value)
+
+    if not all([ca_path, cert_path, key_path]):
+        scheme = params.get(DriverParams.SCHEME.value, "Unknown")
+        role = "Server" if ssl_server else "Client"
+        raise CommError(CommError.BAD_CONFIG, f"{role} certificate parameters are missing for scheme {scheme}")
+
+    if ssl_server:
+        ctx = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
+    else:
+        ctx = ssl.create_default_context()
+
+    # This feature is only supported on 3.7+
+    ctx.minimum_version = ssl.TLSVersion.TLSv1_2
+    ctx.verify_mode = ssl.CERT_REQUIRED
+    ctx.check_hostname = False
+    ctx.load_verify_locations(ca_path)
+    ctx.load_cert_chain(certfile=cert_path, keyfile=key_path)
+
+    return ctx
+
+
+def get_address(params: dict) -> str:
+    host = params.get(DriverParams.HOST.value, "0.0.0.0")
+    port = params.get(DriverParams.PORT.value, 0)
+
+    return f"{host}:{port}"
+
+
+def parse_port_range(entry: Any):
+
+    if isinstance(entry, int):
+        return range(entry, entry + 1)
+
+    parts = entry.split("-")
+    if len(parts) == 1:
+        num = int(parts[0])
+        return range(num, num + 1)
+    lo = int(parts[0]) if parts[0] else LO_PORT
+    hi = int(parts[1]) if parts[1] else HI_PORT
+    return range(lo, hi + 1)
+
+
+def parse_port_list(ranges: Any) -> list:
+    all_ranges = []
+    if isinstance(ranges, list):
+        for r in ranges:
+            all_ranges.append(parse_port_range(r))
+    else:
+        all_ranges.append(parse_port_range(ranges))
+
+    return all_ranges
+
+
+def check_tcp_port(port) -> bool:
+    result = False
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    try:
+        s.bind(("", port))
```

Comment:
This is used in a function to check if port number is used. It needs to check for all interfaces. There is no security risk by checking port.

Reply:
Could we change this line to `s.bind(("127.0.0.1", port))`? will the code scanning still raise this error?

Reply:
Code scanning will be happy but it wouldn't do the job. It will only check the port is not used on that one interface.

Reply:
another thing here, is this method only for TCP?
should we rename it to `try_bind_port` or something more generic?

Note that, we will also need to clean up the codes after these changes are merged.
Since there are duplicate codes that can be reused.

Reply:
This is for TCP only. HTTP/GRPC are all TCP-based protocols.

Reply:
OK, we can write explanation in the code-scanning result